### PR TITLE
Add `GetSnapshotsIT#testAllFeatures`

### PR DESCRIPTION
**Pull request by DaveCTurner**
The features of get-snapshots API are all tested in isolation or small
combinations, but there's no one test which pins down exactly how they
all interact. This commit adds such a test, to verify that any future
optimization work preserves the observable behaviour.

Relates #95345
Relates #104607

### CODE REVIEW

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java [hunk 1]
```diff
@@ -745,4 +762,196 @@ private static GetSnapshotsRequestBuilder baseGetSnapshotsRequest(String[] repoN
...
+        // GetSnapshotsResponse.totalCount whereas snapshots skipped by ?after and ?offset are not
+        final int skippedByFromSortValue;
+        if (randomBoolean()) {
+            final var startingSnapshot = randomFrom(snapshotInfos);
+            getSnapshotsRequest.fromSortValue(switch (sortKey) {
+                case START_TIME -> Long.toString(startingSnapshot.startTime());
+                case NAME -> startingSnapshot.snapshotId().getName();
+                case DURATION -> Long.toString(startingSnapshot.endTime() - startingSnapshot.startTime());
+                case INDICES, SHARDS -> Integer.toString(startingSnapshot.indices().size());
+                case FAILED_SHARDS -> "0";
```
Comment (ywangd) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java:866]: There should be no failed shards. So we effectively do not test it. Should we add a comment?

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java [hunk 2]
```diff
@@ -745,4 +762,196 @@ private static GetSnapshotsRequestBuilder baseGetSnapshotsRequest(String[] repoN
...
+                .size(nextSize)
+                .after(SnapshotSortKey.decodeAfterQueryParam(nextRequestAfter));
+            final GetSnapshotsResponse nextResponse = safeAwait(l -> client().execute(TransportGetSnapshotsAction.TYPE, nextRequest, l));
+
+            assertEquals(
+                selectedSnapshots.stream().skip(nextExpectedOffset).limit(nextSize).map(SnapshotInfo::snapshotId).toList(),
+                nextResponse.getSnapshots().stream().map(SnapshotInfo::snapshotId).toList()
+            );
+            assertEquals(nextSize, nextResponse.getSnapshots().size());
+            assertEquals(selectedSnapshots.size(), nextResponse.totalCount());
```
Comment (ywangd) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java:948]: This is not 
```suggestion
            assertEquals(selectedSnapshots.size() - skippedByFromSortValue, nextResponse.totalCount());
```
?

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java [hunk 3]
```diff
@@ -745,4 +762,196 @@ private static GetSnapshotsRequestBuilder baseGetSnapshotsRequest(String[] repoN
...
+            assertEquals(remaining - nextSize, nextResponse.remaining());
+            assertEquals(nextResponse.remaining() > 0, nextResponse.next() != null);
+
+            nextRequestAfter = nextResponse.next();
+            nextExpectedOffset += nextSize;
+            remaining -= nextSize;
+        }
```
Comment (ywangd) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java:955]: Can we assert `remaining == 0` here?

**Review by ywangd — APPROVED** [review]
LGTM

There is also `slm_policy_filter`. But I guess it is not something of interest in the context of this work?

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java:876]: ++ see 2f2e4854e5ca08c20a193b332dfd54b000ecfe86.

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java:965]: We can, see 58ff48ce1f95df37fb27efc0587d3c6663650807.

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java:958]: Correct. We're not even specifying `?from_sort_value` in this request, so we're seeing every matching snapshot in the total. I think we probably should not subtract these skipped items from the original total either, it seems fairly inconsistent to do what we do today, but this behaviour dates back quite some time so we'd need to discuss any changes. It's noted in this comment:

https://github.com/elastic/elasticsearch/blob/58ff48ce1f95df37fb27efc0587d3c6663650807/server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java#L866-L867

**Comment by DaveCTurner** [comment]
> There is also `slm_policy_filter`

Ah good point, yeah we should cover that in these tests too. I tried a few quick fixes without success, will come back to this in a follow-up.

**Comment by DaveCTurner** [comment]
> I tried a few quick fixes without success, will come back to this in a follow-up.

Spotted the problem, added coverage in c6de0b48c1af2ae559aeb6c2e7cc72d8cfde43f4.

**Pull request merged**